### PR TITLE
cmds/exp/tcpdump: dump traffic on a network device

### PR DESCRIPTION
**Pull request by RiSKeD**
tcpdump takes a network interface as an argument and dumps its traffic. The traffic can be filtered using a BPF filter (see libpcap documentation). The underlying gopacket library uses c-binding, hence we use the go-pcap pkg as well for a pure native go implementation of libpcap (not all filters are supported, I added an explanation as to which are available in the help text of the cmd). What do you think of the concept and example output so far?

 Unfortunately, this introduces a lot of dependencies, i.e. vendor code due to the complexity of the command. Size comparison:
```
 13460100 bytes /tmp/initramfs.linux_amd64.cpio
3823616 bytes /tmp/initramfs.linux_amd64.cpio.xz
15041156 bytes /tmp/initramfs_tcpdump.linux_amd64.cpio
4213248 bytes /tmp/initramfs_tcpdump.linux_amd64.cpio.xz
```
So roughly ~10% of an increase.

Implementation Status: 
- [x] POC & Basic functionalities
- [x] BPF filtering enabled 
- [x] Parsing of network layer:
  - [x] TCP
  - [x] UDP
  - [x] ICMP
- [x] Parsing of application data:
  - [x] DNS
- [x] Flags
  - [x] c
  - [x] h[elp]
  - [x] i[interface]
  - [x] s[napshot-length] 
  - [x] no-promiscuous-mode
  - [x] count
  - [x] D, list-interfaces
  - [x] n[numerical]
  - [x] number, #
  - [x] t, tt, ttt, ttt
  - [x] v[erbose]
  - [x] x
  - [x] xx
  - [x] q
  - [x] A
  - [x] e
  - [x] F
  - [x] nano
- [x] Testing & Coverage

Dump example with TCP and DNS information: 
```
sudo ./tcpdump -i wlp3s0 -c 10
tcpdump: listening on wlp3s0, link-type 1, snapshot length 262144 bytes
18:23:49.778307 IPv4 wlp3s0 l1x0r.fritz.box.47106 > ec2-18-197-249-189.eu-central-1.compute.amazonaws.com.https: Flags [P.], seq 1542549958, ack 1755710844, win 2916, options [NOP,NOP,Timestamps val 640152097], length 55
18:23:49.790192 IPv4 wlp3s0 l1x0r.fritz.box.53736 > fritz.box.domain: 31101+ PTR? 178.178.168.192.in-addr.arpa (46)
18:23:49.791143 IPv4 wlp3s0 ec2-18-197-249-189.eu-central-1.compute.amazonaws.com.443 > l1x0r.fritz.box.47106: Flags [P.], seq 1755710844, ack 1542550013, win 8, options [NOP,NOP,Timestamps val 2718596428], length 57
18:23:49.791168 IPv4 wlp3s0 l1x0r.fritz.box.47106 > ec2-18-197-249-189.eu-central-1.compute.amazonaws.com.https: Flags [.], seq 1542550013, ack 1755710901, win 2916, options [NOP,NOP,Timestamps val 640152109], length 0
18:23:49.792199 IPv4 wlp3s0 wpad.fritz.box.53 > l1x0r.fritz.box.53736: 31101* 1/1/3 PTR l1x0r.fritz.box (161)
18:23:49.792629 IPv4 wlp3s0 l1x0r.fritz.box.59641 > www.fritz.nas.domain: 64500+ PTR? 189.249.197.18.in-addr.arpa (45)
18:23:49.810369 IPv4 wlp3s0 wpad.fritz.box.53 > l1x0r.fritz.box.59641: 64500+PTR ec2-18-197-249-189.eu-central-1.compute.amazonaws.com (112)
18:23:49.811220 IPv4 wlp3s0 l1x0r.fritz.box.60736 > www.fritz.nas.domain: 39193+ PTR? 1.178.168.192.in-addr.arpa (44)
18:23:49.812476 IPv4 wlp3s0 wpad.box.53 > l1x0r.fritz.box.60736: 39193* 11/1/3 PTR fritz.box, PTR www.fritz.box, PTR myfritz.box, PTR www.myfritz.box, PTR fritz.nas, PTR www.fritz.nas, PTR fritz-nas.fritz.box, PTR fritz-nas.box, PTR wpad.box, PTR wpad.fritz.box, PTR 192-168-178-1.fritz.box (366)

10 packets captured
```

### CODE REVIEW

File: cmds/exp/tcpdump/tcp.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+	return strings.TrimRight(opts, ",")
+}
+
+// tcpOptionToString returns a string representation of the TCP option.
+func tcpOptionToString(opt layers.TCPOption) string {
+	switch opt.OptionType {
+	case layers.TCPOptionKindMSS:
+		if len(opt.OptionData) >= 2 {
+			return fmt.Sprintf("%s val %v",
```
Comment (rminnich) [inline cmds/exp/tcpdump/tcp.go:77]: I would prefer that we minimize this multi-line parameter usage. It's not really idiomatic and it makes using tools like grep really hard. Might you consider making code like this one line? We don't have a 80- or 96-char column limit in u-root.

File: cmds/exp/tcpdump/util.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/vishvananda/netlink"
+)
+
+var wellKnownPortsMap = map[string]string{
```
Comment (rminnich) [inline cmds/exp/tcpdump/util.go:14]: can we at some point have a go generate step to create this? My favorite port, 17010, is not in here :-)

**Review by rminnich — COMMENTED** [review]
When I realize what you've done here, given us tcpdump, it's pretty amazing to me how  nicely you've done it, and how little code you needed. Thanks. See my nits. I'm pretty good with it though.

File: cmds/exp/tcpdump/tcp.go (hunk 1 above)
Comment (RiSKeD) [reply cmds/exp/tcpdump/tcp.go]: Yeah, good catch.
 I condensed it into an oneliner for both option types.

File: cmds/exp/tcpdump/util.go (hunk 2 above)
Comment (RiSKeD) [reply cmds/exp/tcpdump/util.go:14]: I haven't looked much into `go generate` myself yet, what advantages would it bring? 
Do we have to write additional tooling for the CI, and so on for this? If so I would like to do this separately from this PR.
I am open to suggestions regarding well-known ports, although I don't know what port 17010 would be ;)

**Comment by rminnich** [comment]
The pci packages uses go generate, it's quite nice actually.

So you would go generate the service number and name pairs. That can come in a later PR, however.

**Review by rminnich — APPROVED** [review]
nothing's perfect on first try, and this is really useful, so ...

**Comment by RiSKeD** [comment]
i see. let's update that in a second round

**Pull request merged**